middlewares/stripPrefix_test.go

```diff
@@ -0,0 +1,153 @@
+package middlewares
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestStripPrefix(t *testing.T) {
+	handlerPath := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprint(w, r.URL.Path)
+	})
+
+	stripServers := make(map[string]*httptest.Server)
+	stripServers["/norules"] = httptest.NewServer(&StripPrefix{
+		Prefixes: []string{},
+		Handler:  handlerPath,
+	})
+
+	stripList := []string{
+		"/",
+		"/stat/",
+		"/stat",
+		"/nomatch",
+	}
+	for _, stripPath := range stripList {
+		stripServers[stripPath] = httptest.NewServer(&StripPrefix{
+			Prefixes: []string{stripPath},
+			Handler:  handlerPath,
+		})
+	}
+
+	suites := []stripPrefixTestSuite{
+		{
+			server: stripServers["/norules"],
+			desc:   "no strip rules",
+			tests: []stripPrefixTestCase{
+				{
+					url:                "/norules",
+					expectedStatusCode: http.StatusNotFound,
+				},
+			},
+		},
+		{
+			server: stripServers["/"],
+			desc:   "strip / to handle wildcard (.*) requests",
+			tests: []stripPrefixTestCase{
+				{
+					url:                "/",
+					expectedStatusCode: http.StatusOK,
+					expectedBody:       "/",
+				},
+			},
+		},
+		{
+			server: stripServers["/stat/"],
+			desc:   "strip /stat/ matching only subpaths",
+			tests: []stripPrefixTestCase{
+				{
+					url:                "/stat",
+					expectedStatusCode: http.StatusNotFound,
+				},
+				{
+					url:                "/stat/",
+					expectedStatusCode: http.StatusOK,
+					expectedBody:       "/",
+				},
+				{
+					url:                "/status",
+					expectedStatusCode: http.StatusNotFound,
+				},
+				{
+					url:                "/stat/us",
+					expectedStatusCode: http.StatusOK,
+					expectedBody:       "/us",
+				},
+			},
+		},
+		{
+			server: stripServers["/stat"],
+			desc:   "strip /stat matching absolute paths and subpaths",
+			tests: []stripPrefixTestCase{
+				{
+					url:                "/stat",
+					expectedStatusCode: http.StatusOK,
+					expectedBody:       "/",
+				},
+				{
+					url:                "/stat/",
+					expectedStatusCode: http.StatusOK,
+					expectedBody:       "/",
+				},
+				{
+					url:                "/status",
+					expectedStatusCode: http.StatusNotFound,
+				},
+				{
+					url:                "/stat/us",
+					expectedStatusCode: http.StatusOK,
+					expectedBody:       "/us",
+				},
+			},
+		},
+		{
+			server: stripServers["/nomatch"],
+			desc:   "no matching strip rules",
+			tests: []stripPrefixTestCase{
+				{
+					url:                "/anyurl",
+					expectedStatusCode: http.StatusNotFound,
+				},
+			},
+		},
+	}
+
+	for _, suite := range suites {
+		suite := suite
+		t.Run(suite.desc, func(t *testing.T) {
+			t.Parallel()
+			defer suite.server.Close()
+
+			for _, test := range suite.tests {
```

Comment:
The given nested suite/test cases approaches comes with a few drawbacks:

- Since the description is attached to the suite, we can't see which test case exactly is failing.
- For the same reason, we cannot execute individual test cases.
- If one test in a suite fails, the entire suite is aborted.

Overall, I also find the nested approach more difficult to understand.

All of this should be solvable by going back to a single, flat hierarchy.